File: objectivec/day4_part2_2019.m
#import <Foundation/Foundation.h>

int main(int argc, const char * argv[]) {
    @autoreleasepool {
        NSString *input = [NSString stringWithContentsOfFile:@"input.txt" encoding:NSUTF8StringEncoding error:nil];
        NSArray *range = [input componentsSeparatedByString:@"-"];
        int lowerBound = [range[0] intValue];
        int upperBound = [range[1] intValue];
        
        int countPart1 = 0;
        int countPart2 = 0;
        
        for (int i = lowerBound; i <= upperBound; i++) {
            NSString *password = [NSString stringWithFormat:@"%d", i];
            
            BOOL hasAdjacentDigits = NO;
            BOOL hasAdjacentPair = NO;
            BOOL neverDecrease = YES;
            
            for (int j = 0; j < password.length - 1; j++) {
                if ([password characterAtIndex:j] == [password characterAtIndex:j+1]) {
                    hasAdjacentDigits = YES;
                    if (j == 0 || [password characterAtIndex:j] != [password characterAtIndex:j-1]) {
                        if (j == password.length - 2 || [password characterAtIndex:j] != [password characterAtIndex:j+2]) {
                            hasAdjacentPair = YES;
                        }
                    }
                }
                if ([password characterAtIndex:j] > [password characterAtIndex:j+1]) {
                    neverDecrease = NO;
                    break;
                }
            }
            
            if (hasAdjacentDigits && neverDecrease) {
                countPart1++;
                if (hasAdjacentPair) {
                    countPart2++;
                }
            }
        }
        
        printf("Part 1: %d\n", countPart1);
        printf("Part 2: %d\n", countPart2);
    }
    return 0;
}

File: swift/day4_part2_2019.swift

import Foundation

let input = try String(contentsOfFile: "input.txt").trimmingCharacters(in: .whitespacesAndNewlines)
let range = input.components(separatedBy: "-").compactMap { Int($0) }
let lowerBound = range[0]
let upperBound = range[1]

func isValidPasswordPart1(_ password: Int) -> Bool {
    let passwordString = String(password)
    var hasAdjacentDigits = false
    var neverDecrease = true
    
    for i in 0..<passwordString.count-1 {
        let currentDigit = Int(String(passwordString[passwordString.index(passwordString.startIndex, offsetBy: i)]))!
        let nextDigit = Int(String(passwordString[passwordString.index(passwordString.startIndex, offsetBy: i+1)]))!
        
        if currentDigit == nextDigit {
            hasAdjacentDigits = true
        }
        
        if currentDigit > nextDigit {
            neverDecrease = false
        }
    }
    
    return hasAdjacentDigits && neverDecrease
}

func isValidPasswordPart2(_ password: Int) -> Bool {
    let passwordString = String(password)
    var hasAdjacentDigits = false
    var neverDecrease = true
    var adjacentCount = 1
    
    for i in 0..<passwordString.count-1 {
        let currentDigit = Int(String(passwordString[passwordString.index(passwordString.startIndex, offsetBy: i)]))!
        let nextDigit = Int(String(passwordString[passwordString.index(passwordString.startIndex, offsetBy: i+1)]))!
        
        if currentDigit == nextDigit {
            adjacentCount += 1
        } else {
            if adjacentCount == 2 {
                hasAdjacentDigits = true
            }
            adjacentCount = 1
        }
        
        if currentDigit > nextDigit {
            neverDecrease = false
        }
    }
    
    if adjacentCount == 2 {
        hasAdjacentDigits = true
    }
    
    return hasAdjacentDigits && neverDecrease
}

var countPart1 = 0
var countPart2 = 0

for password in lowerBound...upperBound {
    if isValidPasswordPart1(password) {
        countPart1 += 1
    }
    
    if isValidPasswordPart2(password) {
        countPart2 += 1
    }
}

print(countPart1)
print(countPart2)
